lxd/storage/drivers/driver_cephfs.go

```diff
@@ -0,0 +1,977 @@
+				_, snapName, _ := shared.ContainerGetParentAndSnapshotName(vol.name)
+
+				cephSnapPath := filepath.Join(dstMountPath, ".snap", snapName)
+				err := os.Mkdir(cephSnapPath, 0711)
+				if err != nil {
+					return err
+				}
+
+				// Make the snapshot path a symlink
+				targetPath := GetVolumeMountPath(d.name, VolumeTypeCustom, fmt.Sprintf("%s/%s", vol.name, snapName))
+				err = os.MkdirAll(filepath.Dir(targetPath), 0711)
+				if err != nil {
+					return err
+				}
+
+				return os.Symlink(cephSnapPath, targetPath)
+			}
+
+			return nil
+		}, op)
+	}, op)
+	if err != nil {
+		return err
+	}
+
+	revertPaths = nil // Don't revert.
+	return nil
+}
+
+func (d *cephfs) DeleteVolume(volType VolumeType, volName string, op *operations.Operation) error {
+	if volType != VolumeTypeCustom {
+		return fmt.Errorf("Volume type not supported")
+	}
+
+	snapshots, err := d.VolumeSnapshots(volType, volName, op)
+	if err != nil {
+		return err
+	}
+
+	if len(snapshots) > 0 {
+		return fmt.Errorf("Cannot remove a volume that has snapshots")
+	}
+
+	volPath := GetVolumeMountPath(d.name, volType, volName)
+
+	// If the volume doesn't exist, then nothing more to do.
+	if !shared.PathExists(volPath) {
+		return nil
+	}
+
+	// Remove the volume from the storage device.
+	err = os.RemoveAll(volPath)
+	if err != nil {
+		return err
+	}
+
+	// Although the volume snapshot directory should already be removed, lets remove it here
+	// to just in case the top-level directory is left.
+	snapshotDir, err := GetVolumeSnapshotDir(d.name, volType, volName)
+	if err != nil {
+		return err
+	}
+
+	err = os.RemoveAll(snapshotDir)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *cephfs) RenameVolume(volType VolumeType, volName string, newName string, op *operations.Operation) error {
+	if volType != VolumeTypeCustom {
+		return fmt.Errorf("Volume type not supported")
+	}
+
+	vol := NewVolume(d, d.name, volType, ContentTypeFS, volName, nil)
+
+	// Create new snapshots directory.
+	snapshotDir, err := GetVolumeSnapshotDir(d.name, volType, newName)
+	if err != nil {
+		return err
+	}
+
+	err = os.MkdirAll(snapshotDir, 0711)
+	if err != nil {
+		return err
+	}
+
+	type volRevert struct {
+		oldPath   string
+		newPath   string
+		isSymlink bool
+	}
+
+	// Create slice to record paths renamed if revert needed later.
+	revertPaths := []volRevert{}
+	defer func() {
+		// Remove any paths rename if we are reverting.
+		for _, vol := range revertPaths {
+			if vol.isSymlink {
+				os.Symlink(vol.oldPath, vol.newPath)
+			} else {
+				os.Rename(vol.newPath, vol.oldPath)
+			}
+		}
+
+		// Remove the new snapshot directory if we are reverting.
+		if len(revertPaths) > 0 {
+			err = os.RemoveAll(snapshotDir)
+		}
+	}()
+
+	// Rename any snapshots of the volume too.
+	snapshots, err := vol.Snapshots(op)
+	if err != nil {
+		return err
+	}
+
+	for _, snapshot := range snapshots {
+		srcSnapshotDir, err := GetVolumeSnapshotDir(d.name, volType, volName)
```

Comment:
That logic doesn't appear specific to the current snapshot so shouldn't be in the loop

Reply:
Fixed